File: apps/admin_backend/src/modules/identity/human-authorization/contracts/field-guards.ts
import {
  OHAC_ERROR_CODE,
  ohacFail,
  ohacOk,
  type OhacResult,
} from './error-codes';

export const DECIMAL_STRING = /^(0|[1-9][0-9]*)$/;
export const LOWERCASE_UUID =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
export const DIGEST = /^sha256:[0-9a-f]{64}$/;

export const asObject = (
  value: unknown,
): Record<string, unknown> | undefined => {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return undefined;
  }
  return value as Record<string, unknown>;
};

export const isNonEmptyString = (value: unknown): value is string =>
  typeof value === 'string' && value.length > 0;

export const isDecimalString = (value: unknown): value is string =>
  typeof value === 'string' && DECIMAL_STRING.test(value);

export const isLowercaseUuid = (value: unknown): value is string =>
  typeof value === 'string' && LOWERCASE_UUID.test(value);

export const isDigest = (value: unknown): value is string =>
  typeof value === 'string' && DIGEST.test(value);

/**
 * Rejects unknown fields before anything else interprets the payload, so a
 * future field cannot be silently ignored by an older contract version.
 */
export const requireExactKeys = (
  object: Record<string, unknown>,
  allowed: readonly string[],
): OhacResult<Record<string, unknown>> => {
  const allowedSet = new Set(allowed);
  for (const key of Object.keys(object)) {
    if (!allowedSet.has(key)) {
      return ohacFail(OHAC_ERROR_CODE.UNKNOWN_FIELD, key);
    }
  }
  for (const key of allowed) {
    if (!Object.prototype.hasOwnProperty.call(object, key)) {
      return ohacFail(OHAC_ERROR_CODE.MISSING_FIELD, key);
    }
  }
  return ohacOk(object);
};

/** Requires an array whose values are already sorted and free of duplicates. */
export const requireSortedUnique = (
  values: readonly string[],
  field: string,
): OhacResult<readonly string[]> => {
  for (let index = 1; index < values.length; index++) {
    if (values[index - 1] >= values[index]) {
      return ohacFail(OHAC_ERROR_CODE.INVALID_FIELD, field);
    }
  }
  return ohacOk(values);
};

File: apps/pos_app/lib/data/models/human_authorization/field_guards.dart
import 'error_codes.dart';

/// Canonical decimal string: `"0"` or a non-zero unsigned decimal without a
/// sign or leading zero. All OHAC integer-like values use this form.
final RegExp decimalStringPattern = RegExp(r'^(0|[1-9][0-9]*)$');
final RegExp lowercaseUuidPattern = RegExp(
  r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$',
);
final RegExp digestPattern = RegExp(r'^sha256:[0-9a-f]{64}$');

/// Returns the value as a JSON object, or `null` when the value is not an
/// object (arrays and scalars are rejected, matching the TS guard).
Map<String, dynamic>? asObject(Object? value) {
  if (value is! Map) return null;
  for (final key in value.keys) {
    if (key is! String) return null;
  }
  return value.cast<String, dynamic>();
}

bool isNonEmptyString(Object? value) => value is String && value.isNotEmpty;

bool isDecimalString(Object? value) =>
    value is String && decimalStringPattern.hasMatch(value);

/// Maximum signed Int64 (exploration.md §7.3 sequenceNumber) as canonical
/// decimal.
const String maxInt64Decimal = '9223372036854775807';

/// Canonical decimal within the signed Int64 range (0..9223372036854775807),
/// no leading zeros. The range is checked lexicographically — digit length,
/// then the max-Int64 bound for 19-digit values — so an over-range or hostile
/// multi-hundred-digit payload is rejected before any BigInt allocation.
bool isInt64DecimalString(Object? value) {
  if (value is! String || !isDecimalString(value)) return false;
  return value.length < maxInt64Decimal.length ||
      (value.length == maxInt64Decimal.length &&
          value.compareTo(maxInt64Decimal) <= 0);
}

bool isLowercaseUuid(Object? value) =>
    value is String && lowercaseUuidPattern.hasMatch(value);

bool isDigest(Object? value) => value is String && digestPattern.hasMatch(value);

/// Rejects unknown fields before anything else interprets the payload, so a
/// future field cannot be silently ignored by an older contract version.
OhacResult<Map<String, dynamic>> requireExactKeys(
  Map<String, dynamic> object,
  List<String> allowed,
) {
  final allowedSet = allowed.toSet();
  for (final key in object.keys) {
    if (!allowedSet.contains(key)) {
      return OhacFailure(OhacError(OhacErrorCode.unknownField, key));
    }
  }
  for (final key in allowed) {
    if (!object.containsKey(key)) {
      return OhacFailure(OhacError(OhacErrorCode.missingField, key));
    }
  }
  return OhacSuccess(object);
}

/// Requires an array whose values are already sorted (UTF-16 code-unit order,
/// identical to JS `<`/`>=` string comparison) and free of duplicates.
OhacResult<List<String>> requireSortedUnique(
  List<String> values,
  String field,
) {
  for (var index = 1; index < values.length; index++) {
    if (values[index - 1].compareTo(values[index]) >= 0) {
      return OhacFailure(OhacError(OhacErrorCode.invalidField, field));
    }
  }
  return OhacSuccess(values);
}
